File: app/src/ultimate/java/com/clockmods/widget/update/WidgetRefreshWorker.java
package com.clockmods.widget.update;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.*;
import com.clockmods.weather.*;
import java.util.concurrent.TimeUnit;
public final class WidgetRefreshWorker extends Worker {
    public static final String PERIODIC="clockmods_widget_weather_refresh";
    public static final String MANUAL="clockmods_widget_weather_manual";
    public WidgetRefreshWorker(@NonNull Context c,@NonNull WorkerParameters p) { super(c,p); }
    @NonNull @Override public Result doWork() {
        Context c=getApplicationContext();
        if(!WidgetRefreshPolicy.shouldSchedule(WidgetUpdateCoordinator.weatherCount(c))) return Result.success();
        WeatherRefreshUseCase.RefreshResult result=new WeatherRefreshUseCase(c).refreshForWidget(getInputData().getBoolean("force",false));
        WidgetUpdateCoordinator.updateWeatherWidgets(c);
        return WidgetRefreshPolicy.shouldRetry(QWeatherConfig.isConfigured(),result==WeatherRefreshUseCase.RefreshResult.TRANSIENT_FAILURE,getRunAttemptCount()) ? Result.retry() : Result.success();
    }
    public static void schedule(Context c,int count) {
        WorkManager manager=WorkManager.getInstance(c);
        if(!WidgetRefreshPolicy.shouldSchedule(count)) { manager.cancelUniqueWork(PERIODIC); manager.cancelUniqueWork(MANUAL); return; }
        manager.enqueueUniquePeriodicWork(PERIODIC,ExistingPeriodicWorkPolicy.UPDATE,
            new PeriodicWorkRequest.Builder(WidgetRefreshWorker.class,30,TimeUnit.MINUTES).setConstraints(network()).build());
    }
    public static void refresh(Context c) {
        if(WidgetUpdateCoordinator.weatherCount(c)==0) return;
        WorkManager.getInstance(c).enqueueUniqueWork(MANUAL,ExistingWorkPolicy.KEEP,
            new OneTimeWorkRequest.Builder(WidgetRefreshWorker.class).setConstraints(network()).setInputData(new Data.Builder().putBoolean("force",true).build()).build());
    }
    private static Constraints network() { return new Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build(); }
}

File: app/src/ultimate/java/com/clockmods/widget/update/WidgetRefreshWorker.kt
package com.clockmods.widget.update

import android.content.Context
import androidx.work.Constraints
import androidx.work.Data
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.clockmods.weather.QWeatherConfig
import com.clockmods.weather.WeatherRefreshUseCase
import java.util.concurrent.TimeUnit

class WidgetRefreshWorker(
    appContext: Context,
    workerParams: WorkerParameters,
) : Worker(appContext, workerParams) {
    override fun doWork(): Result {
        val context = applicationContext
        if (!WidgetRefreshPolicy.shouldSchedule(WidgetUpdateCoordinator.weatherCount(context))) return Result.success()
        val result = WeatherRefreshUseCase(context).refreshForWidget(inputData.getBoolean("force", false))
        WidgetUpdateCoordinator.updateWeatherWidgets(context)
        return if (WidgetRefreshPolicy.shouldRetry(
                QWeatherConfig.isConfigured(),
                result == WeatherRefreshUseCase.RefreshResult.TRANSIENT_FAILURE,
                runAttemptCount,
            )
        ) Result.retry() else Result.success()
    }

    companion object {
        const val PERIODIC = "clockmods_widget_weather_refresh"
        const val MANUAL = "clockmods_widget_weather_manual"

        @JvmStatic fun schedule(context: Context, count: Int) {
            val manager = WorkManager.getInstance(context)
            if (!WidgetRefreshPolicy.shouldSchedule(count)) {
                manager.cancelUniqueWork(PERIODIC)
                manager.cancelUniqueWork(MANUAL)
                return
            }
            manager.enqueueUniquePeriodicWork(
                PERIODIC,
                ExistingPeriodicWorkPolicy.UPDATE,
                PeriodicWorkRequest.Builder(WidgetRefreshWorker::class.java, 30, TimeUnit.MINUTES)
                    .setConstraints(network())
                    .build(),
            )
        }

        @JvmStatic fun refresh(context: Context) {
            if (WidgetUpdateCoordinator.weatherCount(context) == 0) return
            WorkManager.getInstance(context).enqueueUniqueWork(
                MANUAL,
                ExistingWorkPolicy.KEEP,
                OneTimeWorkRequest.Builder(WidgetRefreshWorker::class.java)
                    .setConstraints(network())
                    .setInputData(Data.Builder().putBoolean("force", true).build())
                    .build(),
            )
        }

        private fun network(): Constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()
    }
}
